_posts/2020-07-07-aws-localstack.md

```diff
@@ -0,0 +1,252 @@
+We start in the usual way by creating a profile with any arbitrary credentials.
+```
+aws configure --profile localstack
+```
+Here we create a profile named localstack. You can use any other name. 
+Provide any value you like for aws Access Key and Secret Access Key and a valid AWS region like us-east-1 except leaving them blank. Unlike AWS, LocalStack does not validate these credentials but complains if no profile is set. So far, it is just any other AWS profile which we will use to work with LocalStack.
+
+#### Execute CLI Commands
+With our profile created, we execute CLI commands by passing an additional parameter for overriding the endpoint URL as shown here.
+```
+aws s3 --endpoint-url http://localhost:4566 create-bucket io.pratik.mybucket
+```
+***We create an S3 bucket in LocalStack by overriding the endpoint URL.***
+Next, we execute a regular cloudformation template to create a stack composed of an S3 bucket and a DynamoDB table.
+
+```
+aws cloudformation --endpoint-url http://localhost:4566 create-stack --stack-name samplestack --template-body file://sample.yaml --profile localstack
+```
+
+```
+aws cloudformation --endpoint-url http://localhost:4566 delete-stack --stack-name samplestack --profile localstack
+```
+
+### Execute JUnit Tests
+While running as JUnit, ***LocalStack is started as a docker container at a random port and stopped after all tests have finished execution.*** 
+
+The code snippet is a JUnit Jupiter test used to test a java class to store an object in an S3 bucket.
+
+```java
+@ExtendWith(LocalstackDockerExtension.class)
+@LocalstackDockerProperties(services = { "s3", "sqs" })
+class AwsServiceClientTest {
+
+	private static final Logger logger = Logger.getLogger(AwsServiceClient.class.getName());
+
+	private static final Region region = Region.US_EAST_1;
+	private static final String bucketName = "io.pratik.poc";
+
+	private AwsServiceClient awsServiceClient = null;
+
+	@BeforeEach
+	void setUp() throws Exception {
+		String endpoint = Localstack.INSTANCE.getEndpointS3();
+		awsServiceClient = new AwsServiceClient(endpoint);
+		createBucket();
+	}
+
+	@Test
+	void testStoreInS3() {
+		logger.info("Executing test...");
+		awsServiceClient.storeInS3("image1");
+		assertTrue(keyExistsInBucket(), "Object created");
+	}
+```
+
+
+### Spring Boot Application using S3 and DynamoDB
+
+We will create a simple customer registration application using the popular [Spring Boot](https://spring.io/projects/spring-boot) framework. Our application will have an API that will take a first name, last name, email, mobile, and a profile picture. This API will save the record in DynamoDB, and store the profile picture in an S3 bucket. 
+
+I started by creating a Spring boot rest API using https://start.spring.io with minimum dependencies for web, and Lombok. I created docker-compose.yml. The below snippet shows the environment variable SERVICES set to the name of services (s3 and dynamodb) I need to spin up for my application.
```

Comment:
Switched from "we" to "I" and from future tense to past tense. Please stick with "we" in future or present tense. I edited it in a few places already, but please look through the rest of the text.